### PR TITLE
Support parallel read and write/delete to same key in NonBatchedOpsStressTest

### PR DESCRIPTION
**Pull request by hx235**
**Context:**
Current `NonBatchedOpsStressTest` does not allow multi-thread read (i.e, Get, Iterator) and write (i.e, Put, Merge) or delete to the same key. Every read or write/delete operation will acquire lock (`GetLocksForKeyRange`) on the target key to gain exclusive access to it. This does not align with RocksDB's nature of allowing multi-thread read and write/delete to the same key, that is concurrent threads can issue read/write/delete to RocksDB without external locking. Therefore this is a gap in our testing coverage. 

To close the gap, biggest challenge remains in verifying db value against expected state in presence of parallel read and write/delete. The challenge is due to read/write/delete to the db and read/write to expected state is not within one atomic operation. Therefore we may not know the exact expected state of a certain db read, as by the time we read the expected state for that db read, another write to expected state for another db write to the same key might have changed the expected state.

**Summary:**
Credited to @ajkr's idea, we now solve this challenge by breaking the 32-bits expected value of a key into different parts that can be read and write to in parallel. 

Basically we divide the 32-bits expected value into `value_base` (corresponding to the previous whole 32 bits but now with some shrinking in the value base range we allow), `pending_write` (i.e, whether there is an ongoing concurrent write), `del_counter` (i.e, number of times a value has been deleted, analogous to value_base for write), `pending_delete` (similar to pending_write) and `deleted` (i.e whether a key is deleted). 

Also, we need to use incremental `value_base` instead of random value base as before because we want to control the range of value base a correct db read result can possibly be in presence of parallel read and write. In that way, we can verify the correctness of the read against expected state more easily. This is at the cost of reducing the randomness of the value generated in NonBatchedOpsStressTest we are willing to accept.

(For detailed algorithm of how to use these parts to infer expected state of a key, see the PR)

Misc: hide value_base detail from callers of ExpectedState by abstracting related logics into ExpectedValue class

**Test:**
- Manual test of small number of keys (i.e, high chances of parallel read and write/delete to same key) with equally distributed read/write/deleted for 30 min
```
python3 tools/db_crashtest.py --simple {blackbox|whitebox} --sync_fault_injection=1 --skip_verifydb=0 --continuous_verification_interval=1000 --clear_column_family_one_in=0 --max_key=10 --column_families=1 --threads=32 --readpercent=25 --writepercent=25 --nooverwritepercent=0 --iterpercent=25 --verify_iterator_with_expected_state_one_in=1 --num_iterations=5 --delpercent=15 --delrangepercent=10 --range_deletion_width=5 --use_merge={0|1} --use_put_entity_one_in=0 --use_txn=0 --verify_before_write=0 --user_timestamp_size=0 --compact_files_one_in=1000 --compact_range_one_in=1000 --flush_one_in=1000 --get_property_one_in=1000 --ingest_external_file_one_in=100 --backup_one_in=100 --checkpoint_one_in=100 --approximate_size_one_in=0 --acquire_snapshot_one_in=100 --use_multiget=0 --prefixpercent=0 --get_live_files_one_in=1000 --manual_wal_flush_one_in=1000 --pause_background_one_in=1000 --target_file_size_base=524288 --write_buffer_size=524288 --verify_checksum_one_in=1000 --verify_db_one_in=1000
```
- Rehearsal stress test for normal parameter and aggressive parameter to see if such change can find what existing stress test can find (i.e, no regression in testing capability) 
- [Ongoing]Try to find new bugs with this change that are not found by current NonBatchedOpsStressTest with no parallel read and write/delete to same key

### CODE REVIEW

File: db_stress_tool/expected_state.cc [hunk 1]
```diff
@@ -67,20 +148,21 @@ int ExpectedState::DeleteRange(int cf, int64_t begin_key, int64_t end_key,
 }
 
 bool ExpectedState::Exists(int cf, int64_t key) {
-  // UNKNOWN_SENTINEL counts as exists. That assures a key for which overwrite
+  // Pending states count as "exist". That assures a key for which overwrite
   // is disallowed can't be accidentally added a second time, in which case
```
Comment (hx235) [inline db_stress_tool/expected_state.cc:152]: [Note] can't really think of how the accident can happen ....

File: db_stress_tool/expected_state.cc [hunk 2]
```diff
@@ -3,39 +3,106 @@
...
   std::fill(&Value(cf, 0 /* key */), &Value(cf + 1, 0 /* key */),
-            SharedState::DELETION_SENTINEL);
+            ExpectedState::DEL_MASK);
 }
 
-void ExpectedState::Put(int cf, int64_t key, uint32_t value_base,
-                        bool pending) {
-  if (!pending) {
-    // prevent expected-value update from reordering before Write
-    std::atomic_thread_fence(std::memory_order_release);
```
Comment (hx235) [inline db_stress_tool/expected_state.cc:32]: [Question] Why was this needed previously when the ExpectedState::Put holds lock?

File: db_stress_tool/no_batched_ops_stress.cc [hunk 3]
```diff
@@ -1441,54 +1495,69 @@ class NonBatchedOpsStressTest : public StressTest {
 
   bool VerifyOrSyncValue(int cf, int64_t key, const ReadOptions& /*opts*/,
                          SharedState* shared, const std::string& value_from_db,
-                         const Status& s, bool strict = false) const {
+                         const Status& s) const {
```
Comment (hx235) [inline db_stress_tool/no_batched_ops_stress.cc:1498]: [Note] `strict` is not needed even for existing codes therefore removed now.

File: db_stress_tool/expected_state.cc (hunk 2 above)
Comment (ajkr) [reply db_stress_tool/expected_state.cc:32]: The comment says to prevent reordering within the thread holding the lock. I forgot (like usual) to write the context of why that matters. As you seem to have already realized, it doesn't matter when other threads must grab the lock before reading or writing the same value. But in the case where crash happens after one of the updates but before the other, the order of those updates matters. In particular, if we allowed reordering to happen such that the expected-value was written before the write operation, crash could come between the two. Then recovery would expect a value to be there that was never written.

File: db_stress_tool/expected_state.cc (hunk 1 above)
Comment (ajkr) [reply db_stress_tool/expected_state.cc:152]: It's again a crash recovery case. It could happen that (1) UNKNOWN_SENTINEL written as expected value, (2) write succeeds, (3) crash. Then upon recovery there's an UNKNOWN_SENTINEL for a key that exists in the DB.

Though I wonder if that's still the case after #9897...

File: db_stress_tool/expected_state.cc [hunk 4]
```diff
@@ -67,20 +148,21 @@ int ExpectedState::DeleteRange(int cf, int64_t begin_key, int64_t end_key,
...
 bool ExpectedState::Exists(int cf, int64_t key) {
-  // UNKNOWN_SENTINEL counts as exists. That assures a key for which overwrite
+  // Pending states count as "exist". That assures a key for which overwrite
   // is disallowed can't be accidentally added a second time, in which case
   // SingleDelete wouldn't be able to properly delete the key. It does allow
   // the case where a SingleDelete might be added which covers nothing, but
   // that's not a correctness issue.
   uint32_t expected_value = Value(cf, key).load();
-  return expected_value != SharedState::DELETION_SENTINEL;
+  return PendingWrite(expected_value) || PendingDelete(expected_value) ||
```
Comment (ajkr) [inline db_stress_tool/expected_state.cc:157]: `PendingWrite() || !IsDeleted()` feels like enough - is it not?

**Review by ajkr — APPROVED** [review]
LGTM. There are a few ways to improve readability but overall it's already pretty nice to read - thanks for the great PR!

File: db_stress_tool/expected_state.cc [hunk 5]
```diff
@@ -3,39 +3,106 @@
...
   }
-  Value(cf, key).store(pending ? SharedState::UNKNOWN_SENTINEL : value_base,
-                       std::memory_order_relaxed);
   if (pending) {
-    // prevent Write from reordering before expected-value update
-    std::atomic_thread_fence(std::memory_order_release);
+    SetValuePart(&new_value, PENDING_WRITE_MASK, PENDING_WRITE_MASK);
+    Value(cf, key).store(new_value);
+  } else {
+    SetValuePart(&new_value, VALUE_BASE_MASK, GetValueBase(value));
```
[the remark this answers is not in the document]
Comment (ajkr) [reply db_stress_tool/expected_state.cc:102]: Maybe these should be methods on an `ExpectedValue` class that owns a uint32_t

File: db_stress_tool/no_batched_ops_stress.cc [hunk 6]
```diff
@@ -521,8 +537,9 @@ class NonBatchedOpsStressTest : public StressTest {
           switch (op) {
             case 0:
             case 1: {
-              uint32_t value_base =
-                  thread->rand.Next() % thread->shared->UNKNOWN_SENTINEL;
+              const uint32_t value_base = ExpectedState::NextValueBase(
+                  ExpectedState::GetValueBase(thread->shared->Get(
+                      rand_column_families[0], rand_keys[i])));
```
Comment (ajkr) [inline db_stress_tool/no_batched_ops_stress.cc:542]: It'd be easier to leave this as `thread->rand.Next() % thread->shared->RANDOM_VALUE_BASE_GENERATOR`. I think the counter updates aren't needed since it's only used for Get/MultiGet txn read consistency and then the txn is rolled back

File: db_stress_tool/no_batched_ops_stress.cc [hunk 7]
```diff
@@ -1162,8 +1180,14 @@ class NonBatchedOpsStressTest : public StressTest {
...
-    std::vector<std::unique_ptr<MutexLock>> range_locks =
-        shared->GetLocksForKeyRange(rand_column_family, lb, ub);
+
+    // Parallel read and write to the same key with user timestamp
+    // is not supported, as we might not have the expected value of the key
+    // at some given timestamp
+    if (FLAGS_user_timestamp_size > 0) {
+      std::vector<std::unique_ptr<MutexLock>> range_locks =
+          shared->GetLocksForKeyRange(rand_column_family, lb, ub);
+    }
```
Comment (ajkr) [inline db_stress_tool/no_batched_ops_stress.cc:1190]: Didn't quite understand. What's the difference with `Get()`, which does parallel read and write to the same key? (I think the locking in TestGet() was fully removed.)

**Comment by hx235** [comment]
Thanks for the review - will get back to this next Monday!

**Comment by hx235** [comment]
Mainly rebase by working all my way through related merged PRs (https://github.com/facebook/rocksdb/pull/11133, https://github.com/facebook/rocksdb/pull/11147, https://github.com/facebook/rocksdb/pull/11144, https://github.com/facebook/rocksdb/pull/11228, https://github.com/facebook/rocksdb/pull/11249, https://github.com/facebook/rocksdb/pull/11303) as the rebase is non-trivial (i.e, can't fully trust `git rebase`)

**Comment by hx235** [comment]
Mainly rebase by working all my way through related merged PRs (https://github.com/facebook/rocksdb/pull/11133, https://github.com/facebook/rocksdb/pull/11147, https://github.com/facebook/rocksdb/pull/11144, https://github.com/facebook/rocksdb/pull/11228, https://github.com/facebook/rocksdb/pull/11249, https://github.com/facebook/rocksdb/pull/11303) as the rebase is non-trivial (i.e, can't fully trust `git rebase`)

File: db_stress_tool/expected_state.h [hunk 8]
```diff
@@ -27,6 +27,43 @@ namespace ROCKSDB_NAMESPACE {
...
+  // Return whether value is expected to be deleted from begining till the end
+  // of the read based on `pre_read_expected_value` and
+  // `pre_read_expected_value`.
+  static bool DeletedDuringRead(uint32_t pre_read_expected_value,
+                                uint32_t post_read_expected_value);
+
+  // Return whether value is expected to exist from begining till the end of
+  // the read based on `pre_read_expected_value` and `pre_read_expected_value`.
+  static bool NotDeletedDuringRead(uint32_t pre_read_expected_value,
+                                   uint32_t post_read_expected_value);
```
[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/expected_state.h]: > and MayHaveExisted() for the negation of DeletedDuringRead()?

It seems that we don't really care about MayHaveExisted() but MustHaveExisted(), e.g,  https://github.com/facebook/rocksdb/pull/11058/files#diff-d1176b75e7d3751f59a4387eaa7d94e270df7074d31c6d0874347d1018a98edeR542-R543 cuz that's how we can confidently say something is wrong.

I renamed the functions to be 

```
    static bool MustHaveNotExisted(uint32_t pre_read_expected_value,
                                 uint32_t post_read_expected_value);

  static bool MustHaveExisted(uint32_t pre_read_expected_value,
                              uint32_t post_read_expected_value);
```

Is this better?

File: db_stress_tool/expected_state.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/expected_state.cc]: > It seems clearer if value were not provided at all, and here you use NextValueBase(GetValueBase(new_value))


Are you suggesting we have something like below in `Status TestPut()`

```
    const uint32_t value_base =
        ExpectedValueHelper::NextValueBase(ExpectedValueHelper::GetValueBase(
            shared->Get(rand_column_family, rand_key)));
    const size_t sz = GenerateValue(value_base, value, sizeof(value));
    const Slice v(value, sz);

    shared->Put(rand_column_family, rand_key, true /* pending */); // Right now it's shared->Put(rand_column_family, rand_key value_base, true /* pending */); 
```

I decided to still receive `value_base` in `ExpectedState::Put()` mainly for better semantics with `const size_t sz = GenerateValue(value_base, value, sizeof(value));`, showing explicitly we use same `value_base` across `GenerateValue()` and  `shared->Put()`

I don't have a strong opinion for either way so just want to ensure I understand your suggestion.

File: db_stress_tool/no_batched_ops_stress.cc (hunk 7 above)
Comment (hx235) [reply db_stress_tool/no_batched_ops_stress.cc]: Seems like we don't verify expected state if user timestamp is used in TestGet() https://github.com/facebook/rocksdb/blob/main/db_stress_tool/no_batched_ops_stress.cc#L515 so I just want to align with that. Updating my comments to better reflect my intention.

File: db_stress_tool/expected_state.cc (hunk 1 above)
Comment (hx235) [reply db_stress_tool/expected_state.cc]: As discussed, this comment may no longer be true so I deleted it.

File: db_stress_tool/expected_state.cc (hunk 2 above)
Comment (hx235) [reply db_stress_tool/expected_state.cc:32]: Fixed by adding std::atomic_thread_fence(std::memory_order_release); back as it was

File: db_stress_tool/expected_state.cc (hunk 4 above)
Comment (hx235) [reply db_stress_tool/expected_state.cc]: Fixed

File: db_stress_tool/no_batched_ops_stress.cc (hunk 6 above)
Comment (hx235) [reply db_stress_tool/no_batched_ops_stress.cc]: Fixed

**Comment by hx235** [comment]
Finished rebasing and fixed all the comments, except for [TOCONFIRM]. Will launch stress test again cuz it's been a while since the last passing test.

File: db_stress_tool/expected_state.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (ajkr) [reply db_stress_tool/expected_state.h]: Sure that's fine

File: db_stress_tool/expected_state.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (ajkr) [reply db_stress_tool/expected_state.cc]: > It seems clearer if `value` were not provided at all, and here you use `NextValueBase(GetValueBase(new_value))`

I meant like the way it's done in `ExpectedState::Delete()`: https://github.com/facebook/rocksdb/pull/11058/files#diff-274c08b6c1955e222d9ffb3ebe1ecfbf2e94a0e8bc2417ff2d0b05ba9378d264R133-R135

[the remark this answers is not in the document]
Comment (ajkr) [reply db_stress_tool/expected_state.cc]: BTW, I know this does not solve your problem of how to generate the value for the actual `Put()`, and am still thinking about how to solve that and make things more nicely layered. Right now `ExpectedState` sometimes has its changed decided externally and sometimes decides them internally.

[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/expected_state.cc]: Got you - let me think about it

[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/expected_state.cc]: > Right now ExpectedState sometimes has its changes decided externally and sometimes decides them internally.

I think renaming some function names is good enough: 
```
  const uint32_t value_base = shared->GetNextUnusedValueBase(rand_column_family, rand_key);
  const size_t sz = GenerateValue(value_base, value, sizeof(value));
  const Slice v(value, sz);
  shared->Put(rand_column_family, rand_key, true /* pending */);  // ExpectedState updated to be same as GetNextUnusedValueBase()
```


The only complicated scenario is in recovery cleanup, which is still fine as the `value_base_from_db` should be the GetNextUnusedValueBase() upon crash

```
bool VerifyOrSyncValue(int cf, int64_t key, const ReadOptions& /*opts*/,
                         SharedState* shared, const std::string& value_from_db,
                         std::string msg_prefix, const Status& s) const {
    if (shared->HasVerificationFailedYet()) {
      return false;
    }
    const uint32_t expected_value = shared->Get(cf, key);

    if (ExpectedValueHelper::PendingWrite(expected_value) ||
        ExpectedValueHelper::PendingDelete(expected_value)) {
      if (s.ok()) {
        // Value exists in db, update state to reflect that
        const Slice slice(value_from_db);
        const uint32_t value_base_from_db = GetValueBase(slice);
        shared->Put(cf, key, value_base_from_db, false /* pending */);
      } else if (s.IsNotFound()) {
        // Value doesn't exist in db, update state to reflect that
        shared->SingleDelete(cf, key, false /* pending */);
      }
      return true;
    }
```

[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/expected_state.cc]: Trying out something in the latest commit https://github.com/facebook/rocksdb/pull/11058/commits/66df71deaa7f79ba88236175a9d482cb27fce71e

Conceptually it looks fine to me but will be running stress test on it to see if I miss any edge case (e.g, txn recovery cleanup)

[the remark this answers is not in the document]
Comment (ajkr) [reply db_stress_tool/expected_state.cc]: IMO there is a missed opportunity for encapsulation still. For example, you could make `ExpectedState` functions return `ExpectedValue` class instead of uint32_t, which could have methods like `IncValueBase(bool pending)`, `GetValueBase()`, `GetFinalValueBase()` (this one would apply pending update when there is one), etc.

[the remark this answers is not in the document]
Comment (ajkr) [reply db_stress_tool/expected_state.cc]: > ```
>   const uint32_t value_base = shared->GetNextUnusedValueBase(rand_column_family, rand_key);
>   const size_t sz = GenerateValue(value_base, value, sizeof(value));
>   const Slice v(value, sz);
>   shared->Put(rand_column_family, rand_key, true /* pending */);  // ExpectedState updated to be same as GetNextUnusedValueBase()
> ```

This introduces a coupling between GetNextUnusedValueBase() and Put(), which isn't ideal.

By the suggestion in my last comment it would look like:

```
ExpectedValue ev = shared->Put(rand_column_family, rand_key, true /* pending */);
const size_t sz = GenerateValue(ev.GetFinalValueBase(), value, sizeof(value));
const Slice v(value, sz);
```

**Comment by hx235** [comment]
Just a hack to confirm crash test failure rootcause

File: db_stress_tool/expected_state.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/expected_state.cc]: Fixed

**Comment by hx235** [comment]
Fixed all the comments and now going through a bunch of stress tests.

File: db_stress_tool/expected_state.h [hunk 9]
```diff
@@ -87,7 +272,7 @@ class ExpectedState {
   // member function.
   void Reset();
 
-  std::atomic<uint32_t>* values_;
```
Comment (ajkr) [inline db_stress_tool/expected_state.h:90]: Can we leave it as `std::atomic<uint32_t>* values_`? A `uint32_t` can be wrapped in a `ExpectedValue` before exposing it to the upper layers.

File: db_stress_tool/expected_state.h [hunk 10]
```diff
@@ -42,36 +214,49 @@ class ExpectedState {
   //    guaranteed finished. This is useful for crash-recovery testing when the
   //    process may crash before updating the expected values array.
   //
-  // Requires external locking covering `key` in `cf`.
-  void Put(int cf, int64_t key, uint32_t value_base, bool pending);
+  // Requires external locking covering `key` in `cf` to prevent concurrent
+  // write or delete to the same `key`.
+  ExpectedValue& Put(int cf, int64_t key, bool pending);
```
Comment (ajkr) [inline db_stress_tool/expected_state.h:219]: Can it return an `ExpectedValue` (non-reference)?

Comment (ajkr) [reply db_stress_tool/expected_state.h:219]: Actually I would rather eliminate `bool pending` argument and return a `PendingExpectedValue` that can be committed/aborted:

```
class PendingExpectedValue {
 public:
  explicit PendingExpectedValue(std::atomic<uint32_t>* value_ptr,
                                ExpectedValue orig_value,
                                ExpectedValue final_value)
      : value_ptr_(value_ptr),
        orig_value_(orig_value),
        final_value_(final_value) {}

  void Abort() {
    value_ptr_->store(orig_value_.Read(), std::memory_order_relaxed);
  }

  void Commit() {
    // prevent expected-value update from reordering before Write
    std::atomic_thread_fence(std::memory_order_release);
    value_ptr_->store(final_value_.Read(), std::memory_order_relaxed);
  }

  bool OriginalExists() { return orig_value_.Exists(); }

  uint32_t GetFinalValueBase() { return final_value_.GetValueBase(); }

 private:
  std::atomic<uint32_t>* const value_ptr_;
  const ExpectedValue orig_value_;
  const ExpectedValue final_value_;
};
```

The equivalent of `Put(..., true /* pending */)` would be `PreparePut()`:

```
void ExpectedState::Precommit(int cf, int64_t key, const ExpectedValue& value) {
  Value(cf, key).store(value.Read(), std::memory_order_relaxed);
  // prevent expected-value update from reordering after Write
  std::atomic_thread_fence(std::memory_order_release);
}

PendingExpectedValue ExpectedState::PreparePut(int cf, int64_t key) {
  ExpectedValue expected_value = Load(cf, key);
  const ExpectedValue orig_expected_value = expected_value;
  expected_value.Put(true /* pending */);
  const ExpectedValue pending_expected_value = expected_value;
  expected_value.Put(false /* pending */);
  const ExpectedValue final_expected_value = expected_value;
  Precommit(cf, key, pending_expected_value);
  return PendingExpectedValue(&Value(cf, key), orig_expected_value,
                              final_expected_value);
}
```

The equivalent of `Put(..., false /* pending */)` would be `PendingExpectedValue::Commit()` (shown above). Together, these could be used in `TestPut()`  like this:

```
    auto pending_expected_value =
        shared->PreparePut(rand_column_family, rand_key);
    const uint32_t value_base = pending_expected_value.GetFinalValueBase();
    ...
    DB::Put()
    ...
    pending_expected_value.Commit();
```

And similar for Delete(), SingleDelete(), etc

File: db_stress_tool/expected_state.h [hunk 11]
```diff
@@ -19,9 +19,181 @@
...
+    assert(IsValuePartValid(new_value_part, value_part_mask));
+    ClearValuePart(expected_value, value_part_mask);
+    *expected_value |= new_value_part;
+  }
+
+  static void ClearValuePart(uint32_t* expected_value,
+                             uint32_t value_part_mask) {
+    assert(expected_value);
+    *expected_value &= (~value_part_mask);
+  }
```
Comment (ajkr) [inline db_stress_tool/expected_state.h:195]: I would move at least all of these to `ExpectedValue`. Only a couple (`IsValueBaseValid()` and `IsValuePartValid()`) need to remain static; the rest can become member functions.

File: db_stress_tool/expected_state.h [hunk 12]
```diff
@@ -19,9 +19,181 @@
...
+  // the read based on `pre_read_expected_value` and
+  // `pre_read_expected_value`.
+  static bool MustHaveExisted(uint32_t pre_read_expected_value,
+                              uint32_t post_read_expected_value);
+
+  // Return whether the `value_base` falls within the expected value base
+  static bool InExpectedValueBaseRange(uint32_t value_base,
+                                       uint32_t pre_read_expected_value_base,
+                                       uint32_t post_read_expected_value_base,
+                                       bool post_read_expected_pending_write);
```
Comment (ajkr) [inline db_stress_tool/expected_state.h:76]: I would use the following function signatures:

```
  static bool MustHaveNotExisted(ExpectedValue pre_read_expected_value,
                                 ExpectedValue post_read_expected_value);

  static bool MustHaveExisted(ExpectedValue pre_read_expected_value,
                              ExpectedValue post_read_expected_value);

  static bool InExpectedValueBaseRange(uint32_t value_base,
                                       ExpectedValue pre_read_expected_value,
                                       ExpectedValue post_read_expected_value);
```

**Review by ajkr — CHANGES_REQUESTED** [review]

**Review by ajkr — CHANGES_REQUESTED** [review]
More object design suggestions

File: db_stress_tool/db_stress_shared_state.h [hunk 13]
```diff
@@ -43,11 +43,8 @@ class StressTest;
 // State shared by all concurrent executions of the same benchmark.
 class SharedState {
  public:
-  // indicates a key may have any value (or not be present) as an operation on
-  // it is incomplete.
-  static constexpr uint32_t UNKNOWN_SENTINEL = 0xfffffffe;
-  // indicates a key should definitely be deleted
-  static constexpr uint32_t DELETION_SENTINEL = 0xffffffff;
+  // Used to generate random value base for non-NonBatchedOpsStressTest
+  static constexpr uint32_t RANDOM_VALUE_BASE_GENERATOR = 0xfffffffe;
```
[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/db_stress_shared_state.h:47]: I think they are used in batched_op_stress too https://github.com/facebook/rocksdb/blob/main/db_stress_tool/batched_ops_stress.cc#L35. Also it was suggested to keep it in https://github.com/facebook/rocksdb/pull/11058#discussion_r1123872513 - let me know though if any change in your opinion/thoughts since then :)

[the remark this answers is not in the document]
Comment (ajkr) [reply db_stress_tool/db_stress_shared_state.h:47]: I think you can evaluate whether it's useful or not. The purpose of the linked comment is not about keeping RANDOM_VALUE_BASE_GENERATOR, btw. The purpose is to say we can generate any valid value since it'll be rolled back anyways.

[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/db_stress_shared_state.h:47]: oh okay - I will see whether we can remove or localize it to batched_op_stress

File: db_stress_tool/expected_state.h (hunk 12 above)
Comment (hx235) [reply db_stress_tool/expected_state.h]: Fixed

File: db_stress_tool/expected_state.h (hunk 11 above)
Comment (hx235) [reply db_stress_tool/expected_state.h]: Fixed

File: db_stress_tool/expected_state.h (hunk 10 above)
Comment (hx235) [reply db_stress_tool/expected_state.h]: Fixed

File: db_stress_tool/expected_state.h (hunk 9 above)
Comment (hx235) [reply db_stress_tool/expected_state.h]: Fixed

File: db_stress_tool/db_stress_shared_state.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/db_stress_shared_state.h]: Fixed by localizing it to batched_op_stress

File: db_stress_tool/expected_state.h [hunk 14]
```diff
@@ -42,36 +214,49 @@ class ExpectedState {
...
-  // Requires external locking covering `key` in `cf`.
-  void Put(int cf, int64_t key, uint32_t value_base, bool pending);
+  // Requires external locking covering `key` in `cf` to prevent concurrent
+  // write or delete to the same `key`.
+  ExpectedValue& Put(int cf, int64_t key, bool pending);
 
-  // Requires external locking covering `key` in `cf`.
-  uint32_t Get(int cf, int64_t key) const;
+  // Does not requires external locking.
+  uint32_t Get(int cf, int64_t key);
```
[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/expected_state.h]: Fixed

File: db_stress_tool/expected_state.h [hunk 15]
```diff
@@ -19,9 +19,181 @@
...
+#include "util/autovector.h"
 #include "util/string_util.h"
 
 namespace ROCKSDB_NAMESPACE {
+class ExpectedValue {
+ public:
+  explicit ExpectedValue(std::atomic<uint32_t>* value) : value_(value){};
+  // Requires external locking to prevent concurrent
+  // delete to the same ExpectedValue.
+  bool Exist() const;
```
[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/expected_state.h]: Fixed

File: db_stress_tool/db_stress_shared_state.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (ajkr) [reply db_stress_tool/db_stress_shared_state.h]: Does this mean we can't remove it? It looks arbitrary since 0xfffffffe and 0xffffffff are no longer sentinel values

File: db_stress_tool/expected_state.h (hunk 10 above)
Comment (ajkr) [reply db_stress_tool/expected_state.h]: The reason this became relevant in this PR is because this PR needed to introduce a coupling between `Put(cf, key, true /* pending */)` and `Put(cf, key, false /* pending */)`

**Review by ajkr — APPROVED** [review]
LGTM, thanks for the updates

File: db_stress_tool/db_stress_shared_state.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/db_stress_shared_state.h]: will fix by removing it - I didn't realized batched_op_stress does not need RANDOM_VALUE_BASE_GENERATOR at all (even before my PR)
